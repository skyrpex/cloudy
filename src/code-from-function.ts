import * as cdk from "aws-cdk-lib"
import { runtime } from "@pulumi/pulumi"

import { EsbuildBundling } from "./esbuild-bundling"
import { Construct } from "constructs"

export async function codeFromFunction(function_: (...any: any[]) => any) {
  const tokens: {
    // construct: cdk.Resource
    construct: Construct
    attribute: string
    cfnToken: string
    hash: string
    // hash2: string
  }[] = []
  const serializationContext = {
    // construct: undefined as cdk.Resource | undefined,
    construct: undefined as Construct | undefined,
    attribute: undefined as string | undefined,
    cfnToken: undefined as string | undefined,
  }
  const result = await runtime.serializeFunction(function_, {
    serialize(item) {
      // console.log({
      //   instanceOfResource: item instanceof cdk.Resource,
      //   isConstruct: Construct.isConstruct(item),
      //   isResource:
      //     Construct.isConstruct(item) && cdk.Resource.isResource(item),
      // })
      // if (item && Construct.isConstruct(item) && cdk.Resource.isResource(item)) {
      if (item && Construct.isConstruct(item)) {
        serializationContext.construct = item
      } else if (
        serializationContext.construct &&
        !serializationContext.attribute &&
        typeof item === "string"
      ) {
        serializationContext.attribute = item
      } else if (
        serializationContext.construct &&
        serializationContext.attribute &&
        typeof item === "string" &&
        /^\${Token\[TOKEN\.(\d+)]}$/.test(item)
      ) {
        serializationContext.cfnToken = item
        const { construct, attribute, cfnToken } = serializationContext
        if (construct && attribute && cfnToken) {
          // console.log({ attribute, cfnToken })
          tokens.push({
            construct,
            attribute,
            cfnToken,
            // hash: `${cdk.Names.uniqueId(construct)}${attribute}_${
            //   tokens.length
            // }`,
            hash: `${cdk.Names.uniqueId(construct)}${attribute}`,
          })
        }
        serializationContext.attribute = undefined
        serializationContext.cfnToken = undefined
      }
      return true
    },
  })

  let sourceCode = result.text

  for (const token of tokens) {
    sourceCode = sourceCode.replace(
      `"${token.cfnToken}"`,
      `process.env["${token.hash}"]`,
    )
  }

  const code = EsbuildBundling.bundle({
    stdin: {
      contents: sourceCode,
      resolveDir: process.cwd(),
    },
    // define
    // external
  })

  return {
    code,
    tokens,
  }
}
