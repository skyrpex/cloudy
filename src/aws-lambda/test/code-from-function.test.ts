import { expect, test } from "vitest";

import { codeFromFunction } from "../code-from-function.js";

test("captures symbols correctly", async () => {
  const symbol = Symbol();
  const code = await codeFromFunction(() => {
    console.log({ symbol });
  });
  expect(code.sourceCode).toMatchSnapshot();
});
