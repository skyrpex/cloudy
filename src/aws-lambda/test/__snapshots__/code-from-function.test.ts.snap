// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`captures symbols correctly 1`] = `
"exports.handler = __f0;

var __symbol = Symbol();

function __f0() {
  return (function() {
    with({ symbol: __symbol, this: undefined, arguments: undefined }) {

return () => {
    console.log({ symbol });
  };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
"
`;
